packages/cozy-harvest-lib/src/components/TriggerLauncher.jsx

```diff
@@ -0,0 +1,150 @@
+ * * launch: Callback to launch the trigger
+ * * running: Boolean which indicates if the trigger is running
+ *
+ * ### Example
+ * ```js
+ * <TriggerLauncher trigger={trigger}>
+ *   {(launch, running) => (
+ *     <Button onClick={launch} disabled={running}
+ *    )}
+ * </TriggerLauncher>
+ * ```
+ *
+ * TODO: inject other props to indicates status (for example: loginSucceed),
+ * and use callbacks like `onLoginSuccess` to make this component usable in
+ * TriggerManager for example
+ */
+export class TriggerLauncher extends Component {
+  state = {
+    status: IDLE
+  }
+
+  constructor(props, context) {
+    super(props, context)
+
+    this.launch = this.launch.bind(this)
+    this.handleTwoFAModalDismiss = this.handleTwoFAModalDismiss.bind(this)
+    this.handle2FACode = this.handle2FACode.bind(this)
+    this.handleError = this.handleError.bind(this)
+    this.handleSuccess = this.handleSuccess.bind(this)
+    this.handleLoginSuccess = this.handleLoginSuccess.bind(this)
+    this.handleTwoFARequest = this.handleTwoFARequest.bind(this)
+  }
+
+  handleTwoFAModalDismiss() {
+    // TODO: Make the modale not closable, or offer possibility to re-open it.
+    this.setState({ showTwoFAModal: false })
+  }
+
+  handleError() {
+    this.setState({ status: ERRORED, showTwoFAModal: false })
+  }
+
+  handleSuccess() {
+    this.setState({ status: SUCCESS, showTwoFAModal: false })
+  }
+
+  handleLoginSuccess() {
+    this.setState({ status: LOGIN_SUCCESS })
+  }
+
+  handleTwoFARequest() {
+    this.setState({ status: TWO_FA_REQUEST, showTwoFAModal: true })
+  }
+
+  handleTwoFAMismatch() {
+    this.setState({ status: TWO_FA_MISMATCH, showTwoFAModal: true })
+  }
+
+  handle2FACode(code) {
+    this.setState({ status: RUNNING_TWOFA })
+    this.konnectorJob.sendTwoFACode(code)
+  }
+
+  async launch() {
+    const { trigger } = this.props
+    const { client } = this.context
+
+    this.setState({ status: PENDING })
+
+    this.konnectorJob = new KonnectorJob(client, trigger)
+    this.konnectorJob
+      .on('error', this.handleError)
+      .on('loginSuccess', this.handleLoginSuccess)
+      .on('success', this.handleSuccess)
+      .on('twoFARequest', this.handleTwoFARequest)
+      .on('twoFAMismatch', this.handleTwoFAMismatch)
+    this.konnectorJob.launch()
+  }
+
+  render() {
+    const { showTwoFAModal, status } = this.state
+    const { account, children, konnector, submitting } = this.props
+
+    return (
+      <div>
+        {children({
+          launch: this.launch,
+          running: ![ERRORED, IDLE, SUCCESS].includes(status) || submitting
+        })}
+        {showTwoFAModal && (
+          <TwoFAForm
+            account={account}
+            konnector={konnector}
+            dismissAction={this.handleTwoFAModalDismiss}
+            handleSubmitTwoFACode={this.handle2FACode}
+            submitting={status === RUNNING_TWOFA}
+            retryAsked={status === TWO_FA_MISMATCH}
```

Comment:
I was asking myself the same kind of question during development. What made me choose this solutions was to avoid passing a complex object like a KonnectorJob as prop. But I am still not sure this was a good solution.

How I see it if we pass the konnectorJob to TwoFAModal:
* TriggerLauncher must still listen to `twoFARequest` event to show 2FA modal
* TwoFAModal only listen to `twoFAMismatch` event

Do you agree ?

Reply:
Now the TriggerLauncher just use KonnectorJob from props with https://github.com/cozy/cozy-libs/pull/415/commits/fee5d4745535bc71fb6badaf248b2b3fcfb847d3